src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepoContentsCache.java

```diff
@@ -128,4 +165,90 @@ private static String getNextCounterInDir(Path entryDir) throws IOException {
+   * more than {@code maxAge} ago, with an idle delay of {@code idleDelay}.
+   */
+  public IdleTask createGcIdleTask(Duration maxAge, Duration idleDelay) {
+    Preconditions.checkState(path != null);
+    return new IdleTask() {
+      @Override
+      public String displayName() {
+        return "Repo contents cache garbage collection";
+      }
+
+      @Override
+      public Duration delay() {
+        return idleDelay;
+      }
+
+      @Override
+      public void run() throws InterruptedException, IdleTaskException {
+        try {
+          try (var lock = FileSystemLock.tryGet(path.getRelative(LOCK_PATH), LockMode.EXCLUSIVE)) {
+            // If we can't grab the lock, abort GC. Someone will come along later.
+            runGc(maxAge);
+          }
+        } catch (IOException e) {
+          throw new IdleTaskException(e);
+        }
+      }
+    };
+  }
+
+  private void runGc(Duration maxAge) throws InterruptedException, IOException {
+    Preconditions.checkState(path != null);
+
+    path.setLastModifiedTime(Path.NOW_SENTINEL_TIME);
+    Instant cutoff = Instant.ofEpochMilli(path.getLastModifiedTime()).minus(maxAge);
+    // Since deleting entire directories could take a bit of time, we create a "trash" directory
+    // where we move the garbage directories to (which should be very fast). Then we can delete
+    // this trash directory altogether at the end. This makes the GC process safe against being
+    // interrupted in the middle (any undeleted trash will get deleted by the next GC).
+    // Also be sure to name this trashDir something that couldn't ever be a predeclared inputs hash
+    // (starting with an underscore should suffice).
+    Path trashDir = path.getChild("_trash");
+    trashDir.createDirectoryAndParents();
+    // To make sure we don't get a name clash when moving things into the trash, we use the current
+    // time as an entry prefix and keep a counter. If the user is messing around with the system
+    // clock, then $DEITY help them.
+    String trashEntryPrefix = path.getLastModifiedTime() + ".";
+    int trashCounter = 1;
+
+    for (Path entryDir : path.getDirectoryEntries()) {
+      if (!entryDir.isDirectory() || entryDir.equals(trashDir)) {
+        continue;
+      }
+      for (Path recordedInputsFile : entryDir.getDirectoryEntries()) {
+        if (!recordedInputsFile.getBaseName().endsWith(RECORDED_INPUTS_SUFFIX)) {
+          continue;
+        }
+        if (Thread.interrupted()) {
+          throw new InterruptedException();
+        }
+
+        if (Instant.ofEpochMilli(recordedInputsFile.getLastModifiedTime()).isBefore(cutoff)) {
+          // Sorry buddy, you're out.
+          recordedInputsFile.delete();
+          var repoDir = CandidateRepo.fromRecordedInputsFile(recordedInputsFile).contentsDir;
+          repoDir.renameTo(trashDir.getChild(trashEntryPrefix + (trashCounter++)));
```

Comment:
Why not use a UUID? `$DEITY` may be unset ;⁠-⁠)

Reply:
Done. I guess the chance of a user messing around with the system clock, no matter how low, is probably still much higher than UUID clashes :)